ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/recovery/WorkerService.java

```diff
@@ -63,7 +63,7 @@ public void doWork (InputStream is, OutputStream os) throws IOException
 	{
 	    String request = in.readLine();
 
-        if (request.equals("PING"))
+        if ("PING".equals(request))
```

Comment:
Would it be possible to add a test for this please? I think the description of https://issues.redhat.com/browse/JBTM-3257 indicates that the scan would have started but if I understand where this condition can occur it is before the instruction is read. I wonder why an IOException is not coming out of in.readLine?

Reply:
ok. investigation was not easy but I assume I got the reason. see my comment below

Reply:
Sorry to go back in time so far by when you refer to a comment to look at below please can you indicate the one I should look at?

Reply:
@tomjenkinson I meant the one where I summarized my effort. It's this one: https://github.com/jbosstm/narayana/pull/1561#issuecomment-589639235
But as you went through the code already I assume there is not many new information in that comment.

Reply:
Thanks @ochaloup I am wondering which of the tests will make the variable request be equal to null? Is that scenario able to be created in the unit tests using WorkerService? Something like create the PrintWriter client side and then just close the socket and somehow verify that the WorkerService did not NPE when it got a null https://github.com/jbosstm/narayana/pull/1561/files#diff-133ca2f8c1781ec3f48d3b73b2060a23L64

Reply:
@ochaloup Testing getting a null on the request is the key thing I am trying to get to for JBTM-3257. Which of the tests results in that?

I was thinking that a test could open the socket and then close the socket (without sending anything) and it might result in `request` from `String request = in.readLine();` being null?

If you do need to send "SCAN" as described in https://issues.redhat.com/browse/JBTM-3257 then perhaps rather than `toServer.println` you could use https://docs.oracle.com/javase/7/docs/api/java/io/PrintWriter.html#print(java.lang.String) rather than println but I don't know of a way to abruptly kill the socket in Java so perhaps that is not possible / predictable in a platform neutral manner?

Reply:
@tomjenkinson 
test which you ask for is here: https://github.com/ochaloup/narayana/commit/65e56fc545b2d6d85579533c03a923a266a39642
...but it's not possible to be really tested. The trouble is that NPE is thrown but NPE is left just as an error in test log. NPE is at a thread which just disappears and does not influence any functionality.

I didn't add the test into the test suite as it's not easily possible to find the NPE happens or not. And from my perspective it's not an important test. The functionality is not harmed. On socket timeout the client does not wait for any information. So the NPE is only printed to log. While no information is delivered wrongly.

Reply:
I think this is the part I am not sure of most. What to do about that test as that is the part that really covers testing of JBTM-3257 IIUC. I will consider further.

Reply:
@ochaloup please can you take a look at https://github.com/ochaloup/narayana/pull/8 ?

Reply:
@tomjenkinson wow, that's nice. Thanks! I will add the test to PR and I will try to reformat the commits. I hope in a good order :-)